src/com/mostc/pftt/util/HostEnvUtil.java

```diff
@@ -250,7 +495,131 @@ public static boolean regQueryAdd(ConsoleManager cm, AHost host, String key, Str
+	}
+	
+	private static void downloadMySQL5(FileSystemScenario fs, LocalHost host, LocalConsoleManager cm) {
+
+		if(!fs.exists(Exe_Mysql_5_7_mysqld))
+		{
+			// download MySQL and unzip 
+			createDirectoryIfNotExists(fs, cm, Dir_Mysql);
+			DownloadUtil.downloadAndUnzip(cm, host, Link_Mysql_Win32_5_7_25, Dir_Mysql);
+		}
+		else
+		{
+			cm.println(EPrintType.CLUE, HostEnvUtil.class, "MySQL 5.7.25 is already downloaded.");
+		}
+	}
+	
+	private static void downloadVCRuntimeByBuild(FileSystemScenario fs, LocalHost host, LocalConsoleManager cm,
+			PhpBuild build)
+		throws IllegalStateException, IOException, Exception {
+		if (!host.isWindows()) {
+			return;
+		}
+
+		String system_dir = host.getSystemRoot();
+		switch (build.getVersionBranch(cm, host)) {
+		case PHP_5_3:
+		case PHP_5_4:
+			downloadVCRuntime9(fs, host, cm);
+			break;
+		case PHP_5_5:
+		case PHP_5_6:
+			downloadVCRuntime(fs, cm, "VC10 x86", Link_VC10_Redist_X86, File_VC10_Redist_X86, system_dir + Sys_Dll_VC10_Redist_X86);
+			downloadVCRuntime(fs, cm, "VC11 x86", Link_VC11_Redist_X86, File_VC11_Redist_X86, system_dir + Sys_Dll_VC11_Redist_X86);
+			if (build.isX64()) {
+				downloadVCRuntime(fs, cm, "VC10 x64", Link_VC10_Redist_X64, File_VC10_Redist_X64, system_dir + Sys_Dll_VC10_Redist_X64);
+				downloadVCRuntime(fs, cm, "VC11 x64", Link_VC11_Redist_X64, File_VC11_Redist_X64, system_dir + Sys_Dll_VC11_Redist_X64);
+			}
+			break;
+		case PHP_7_0:
+		case PHP_7_1:
+			downloadVCRuntime(fs, cm, "VC14 x86", Link_VC14_Redist_X86, File_VC14_Redist_X86, system_dir + Sys_Dll_VC14_Redist_X86);
+			if (build.isX64()) {
+				downloadVCRuntime(fs, cm, "VC14 x64", Link_VC14_Redist_X64, File_VC14_Redist_X64, system_dir + Sys_Dll_VC14_Redist_X64);
+			}
+			break;
+		case PHP_7_2:
+		case PHP_7_3:
+		case PHP_Master:
+		default:
+			downloadVC15Runtime(fs, cm, "VC15 x86", Link_VC15_Redist_X86, File_VC15_Redist_X86, system_dir + Sys_Dll_VC15_Redist_X86);
+			if (build.isX64()) {
+				downloadVC15Runtime(fs, cm, "VC15 x64", Link_VC15_Redist_X64, File_VC15_Redist_X64, system_dir + Sys_Dll_VC15_Redist_X64);
+			}
+			break;
+		} // end switch
+	}
 			
+	private static void downloadVCRuntime9(FileSystemScenario fs, LocalHost host, LocalConsoleManager cm)
+	{
+		if (installedVCRT9(host)) {
+			cm.println(EPrintType.CLUE, HostEnvUtil.class, "VC9 Runtime already installed, skip downloading.");
+		}
+		else
+		{
+			String pftt_dir = LocalHost.getLocalPfttDir();
+			downloadFile(fs, cm, "VC9 Runtime", Link_VC9_Redist_X86, pftt_dir + File_VC9_Redist_X86);
+		}
+	}
+		
+	private static void downloadVC15Runtime(FileSystemScenario fs, LocalConsoleManager cm,
+			String name, String remote_url, String installer_file, String dll_file)
+	{
+		if(installedVCRT15(fs, dll_file))
+		{
+			cm.println(EPrintType.CLUE, HostEnvUtil.class, name + " Runtime already installed, skip downloading.");
+		}
+		else
+		{
+			downloadFile(fs, cm, name, remote_url, installer_file);
```

Comment:
Shouldn't this check the return value (as well as at other places where downloadFile is used)? Or it's covered by an exception?

Thanks.

Reply:
Currently there is no exception, instead there will be some "error"/"warning" message printed. Maybe exceptions work better here.

Reply:
Either way. Just if the download has failed, it's a blocking issue for the PFTT setup process. The message is good, but FPTT should stop at that point then as further operations might base on that download and thus would be broken, too.

Thanks.